pkg/api/meta/restmapper.go

```diff
@@ -468,91 +468,59 @@ func (o resourceByPreferredGroupVersion) Less(i, j int) bool {
 // RESTClient should use to operate on the provided group/kind in order of versions. If a version search
 // order is not provided, the search order provided to DefaultRESTMapper will be used to resolve which
 // version should be used to access the named group/kind.
-// TODO: consider refactoring to use RESTMappings in a way that preserves version ordering and preference
 func (m *DefaultRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*RESTMapping, error) {
+	mappings, err := m.commonRESTMappings(gk, versions...)
+	if err != nil {
+		return nil, err
+	}
+	if len(mappings) == 0 {
+		return nil, &NoKindMatchError{PartialKind: gk.WithVersion("")}
+	}
+	return mappings[0], nil
+}
+
+// RESTMappings returns the RESTMappings for the provided group kind in a rough internal preferred order. If no
+// kind is found it will return a NoResourceMatchError.
+func (m *DefaultRESTMapper) RESTMappings(gk schema.GroupKind) ([]*RESTMapping, error) {
+	mappings, err := m.commonRESTMappings(gk, "")
+	if err != nil || len(mappings) == 0 {
+		return nil, err
+	}
+	return mappings, nil
+}
+
+// commonRESTMappings returns RESTMappings for the provided group kind. If a version search order
+// is not provided, the search order provided to DefaultRESTMapper will be used.
+func (m *DefaultRESTMapper) commonRESTMappings(gk schema.GroupKind, versions ...string) ([]*RESTMapping, error) {
+	var mappings []*RESTMapping
+	var potentialGVK []schema.GroupVersionKind
+	var hadVersion bool
+
 	// Pick an appropriate version
-	var gvk *schema.GroupVersionKind
-	hadVersion := false
 	for _, version := range versions {
 		if len(version) == 0 || version == runtime.APIVersionInternal {
 			continue
 		}
-
 		currGVK := gk.WithVersion(version)
 		hadVersion = true
 		if _, ok := m.kindToPluralResource[currGVK]; ok {
-			gvk = &currGVK
+			potentialGVK = append(potentialGVK, currGVK)
 			break
 		}
 	}
 	// Use the default preferred versions
-	if !hadVersion && (gvk == nil) {
+	if !hadVersion && len(potentialGVK) == 0 {
 		for _, gv := range m.defaultGroupVersions {
 			if gv.Group != gk.Group {
 				continue
 			}
-
-			currGVK := gk.WithVersion(gv.Version)
-			if _, ok := m.kindToPluralResource[currGVK]; ok {
-				gvk = &currGVK
-				break
-			}
-		}
-	}
-	if gvk == nil {
-		return nil, &NoKindMatchError{PartialKind: gk.WithVersion("")}
```

Comment:
Yes - the code will loop over all m.defaultGroupVersion - I can extend commonRESTMappings method by a parameter and stop on first match